camayoc/tests/qcs/cli/test_sources.py

```diff
@@ -0,0 +1,879 @@
+# coding=utf-8
+"""Tests for ``qpc source`` commands.
+
+:caseautomation: automated
+:casecomponent: cli
+:caseimportance: high
+:caselevel: integration
+:requirement: Sonar
+:testtype: functional
+:upstream: yes
+"""
+import json
+import random
+from io import BytesIO
+
+import pexpect
+import pytest
+
+from camayoc import utils
+from camayoc.constants import CONNECTION_PASSWORD_INPUT
+from camayoc.tests.qcs.cli.utils import cred_add, source_show
+
+
+@pytest.mark.parametrize('hosts', ('127.0.0.1', '192.168.0.0/24'))
+def test_add_with_cred_hosts(isolated_filesystem, qpc_server_config, hosts):
+    """Add a source with cred and hosts.
+
+    :id: 665d76d5-db4c-4e2e-869d-3f97fb0ef878
+    :description: Add a source entry providing the ``--name``, ``--cred`` and
+        ``--hosts`` options.
+    :steps: Run ``qpc source add --name <name> --cred <cred>
+        --hosts <hosts> --type <type>``
+    :expectedresults: A new source entry is created with the data provided as
+        input.
+    """
+    cred_name = utils.uuid4()
+    name = utils.uuid4()
+    cred_add(
+        {
+            'name': cred_name,
+            'username': utils.uuid4(),
+            'password': None,
+        },
+        [(CONNECTION_PASSWORD_INPUT, utils.uuid4())],
+    )
+
+    qpc_source_add = pexpect.spawn(
+        'qpc source add --name {} --cred {} --hosts {} --type network'
+        .format(name, cred_name, hosts)
+    )
+    assert qpc_source_add.expect('Source "{}" was added'.format(name)) == 0
+    assert qpc_source_add.expect(pexpect.EOF) == 0
+    qpc_source_add.close()
+    assert qpc_source_add.exitstatus == 0
+
+    if hosts.endswith('0/24'):
+        hosts = hosts.replace('0/24', '\[0:255\]')
+    source_show(
+        {'name': name},
+        '{{\r\n'
+        '    "credentials": \[\r\n'
+        '        {{\r\n'
+        '            "id": \\d+,\r\n'
+        '            "name": "{}"\r\n'
+        '        }}\r\n'
+        '    \],\r\n'
+        '    "hosts": \[\r\n'
+        '        "{}"\r\n'
+        '    \],\r\n'
+        '    "id": \\d+,\r\n'
+        '    "name": "{}",\r\n'
+        '    "port": 22,\r\n'
+        '    "source_type": "network"\r\n'
+        '}}\r\n'
+        .format(cred_name, hosts, name)
+    )
+
```

Comment:
Don't see any tests regarding the "type" of the source. Examples: using the wrong type credential, invalid type for source, changing type of source, or not specifying the type. If you don't want to implement I would suggest at least a stub or issue to remind yourself

Reply:
This is the missing part, the approach I've used on these changes was to grab what we have for `rho profile` and use that as a initial coverage. The current tests are pretty much equivalent to `rho profile` ones.

That said, type is hard coded on this iteration, my next steps will be see if I can parametrize type for the current tests and add some specific to the type option.

I mean, I am just clarifying what I had in mind since you have a really valid point here. Thank you for bringing it up. I will request another round of review after adding those improvements and then you can see if I am missing something. You have a far more experience with the source API than I do so your input is valuable.